walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,139 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {  
+    // edge case: if there are no attendees the whole day is returned  
+    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();      
+    if (requestAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // edge case: if the requested meeting is longer than a day there are no options
+    long requestDuration = request.getDuration();
+    if (requestDuration >= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();  
+    }
+    ArrayList<TimeRange> alreadyBookedTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> alreadyBookedTimesWOptional = new ArrayList<TimeRange>();
+    alreadyBookedTimes = getAlreadyBookedTimes(events, request);
+    alreadyBookedTimesWOptional = getAlreadyBookedTimesWithOptional(events, request, alreadyBookedTimes);
```

Comment:
nit: Throughout this file, you can simplify some of the declarations. For example: 

```ArrayList<TimeRange> alreadyBookedTimes = getAlreadyBookedTimes(events, request);```

Reply:
Addressed in "simplify variable declarations"